pandas/core/internals/managers.py

```diff
@@ -1248,18 +1185,14 @@ def value_getitem(placement):
             self.blocks = tuple(
                 blk for i, blk in enumerate(self.blocks) if i not in set(removed_blknos)
             )
-            if self.refs is not None:
-                self.refs = [
-                    ref
-                    for i, ref in enumerate(self.refs)
-                    if i not in set(removed_blknos)
-                ]
 
         if unfit_val_locs:
             unfit_idxr = np.concatenate(unfit_mgr_locs)
             unfit_count = len(unfit_idxr)
 
             new_blocks: list[Block] = []
+            # TODO(CoW) is this always correct to assume that the new_blocks
+            # are not referencing anything else?
```

Comment:
There are some tests in `copy_view/test_setitem.py` that checks we do copy the array-like values we are setting (not sure how complete they are though, given this is a beast of a method). And so because we copy upfront, that currently is fine regarding CoW (for setting full columns, which I think is what `iset` is used for?). Although a future optimization we can add is that we can actually delay the copy for Series/DataFrame being set as new column(s). And then we will have to passthrough its refs somehow to this method.

EDIT: Ah, I see it was my own comment that was moved ;)

Reply:
Yep :) This is on my todo list as well, just moved it to a better place after the modification